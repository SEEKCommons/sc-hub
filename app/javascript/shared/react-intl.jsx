/**
 * @flow
 */

import * as React from 'react'
import { FormattedMessage } from 'react-intl'
import { intersperse } from 'ramda'

type Props = {
  list: React.Node[],
  truncate?: { after: number, with: React.Node },
}
export function FormattedList({ list, truncate }: Props): React.Node {
  if (truncate != null && list.length > truncate.after) {
    const firstN = list.slice(0, truncate.after)
    return (
      <span>
        {intersperse(<FormattedMessage id="helpers.list.middle" />, firstN)}
        <FormattedMessage id="helpers.list.middle" />
        {truncate.with}
      </span>
    )
  }

  switch (list.length) {
    case 0:
      return <span />

    case 1:
      return <span>{list[0]}</span>

    case 2:
      return (
        <span>
          {list[0]}
          <FormattedMessage id="helpers.list.two" />
          {list[1]}
        </span>
      )

    default: {
      const start = list[0]
      const middle = list.slice(1, -1)
      const end = list[list.length - 1]
      return (
        <span>
          {start}
          <FormattedMessage id="helpers.list.start" />
          {intersperse(<FormattedMessage id="helpers.list.middle" />, middle)}
          <FormattedMessage id="helpers.list.end" />
          {end}
        </span>
      )
    }
  }
}
