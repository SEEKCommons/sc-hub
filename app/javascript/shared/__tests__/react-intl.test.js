import { FormattedList } from '../react-intl'

import React from 'react'
import renderer from 'react-test-renderer'

jest.mock('react-intl', () => ({
  FormattedMessage: jest.fn(props => (
    <div data-component="FormattedMessage" {...props} />
  )),
}))

describe('FormattedList', () => {
  it('works with zero items', () => {
    const component = renderer.create(<FormattedList list={[]} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('works with one item', () => {
    const component = renderer.create(<FormattedList list={['1']} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('works with two items', () => {
    const component = renderer.create(<FormattedList list={['1', '2']} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('works with three items', () => {
    const component = renderer.create(<FormattedList list={['1', '2', '3']} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('can truncate when the list is too long', () => {
    const component = renderer.create(
      <FormattedList
        list={['1', '2', '3', '4', '5', '6']}
        truncate={{ after: 2, with: 'etc.' }}
      />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('doesnâ€™t truncate if the list is short enough', () => {
    const component = renderer.create(
      <FormattedList
        list={['1', '2', '3', '4']}
        truncate={{ after: 5, with: 'etc.' }}
      />
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
