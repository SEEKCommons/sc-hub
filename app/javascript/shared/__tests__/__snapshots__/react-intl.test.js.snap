// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedList can truncate when the list is too long 1`] = `
<span>
  1
  <div
    data-component="FormattedMessage"
    id="helpers.list.middle"
  />
  2
  <div
    data-component="FormattedMessage"
    id="helpers.list.middle"
  />
  etc.
</span>
`;

exports[`FormattedList doesnâ€™t truncate if the list is short enough 1`] = `
<span>
  1
  <div
    data-component="FormattedMessage"
    id="helpers.list.start"
  />
  2
  <div
    data-component="FormattedMessage"
    id="helpers.list.middle"
  />
  3
  <div
    data-component="FormattedMessage"
    id="helpers.list.end"
  />
  4
</span>
`;

exports[`FormattedList works with one item 1`] = `
<span>
  1
</span>
`;

exports[`FormattedList works with three items 1`] = `
<span>
  1
  <div
    data-component="FormattedMessage"
    id="helpers.list.start"
  />
  2
  <div
    data-component="FormattedMessage"
    id="helpers.list.end"
  />
  3
</span>
`;

exports[`FormattedList works with two items 1`] = `
<span>
  1
  <div
    data-component="FormattedMessage"
    id="helpers.list.two"
  />
  2
</span>
`;

exports[`FormattedList works with zero items 1`] = `<span />`;
